### PR TITLE
Enabling Subclassing of `Reflect.Loader` (major reorganization)

### PR DESCRIPTION
**Pull request by caridy**
_[compiled spec from this PR](https://rawgit.com/caridy/d189efc249d40f2242af/raw/490d9586e646b5fa48b7919d87f68e8e9204c941/loader-spec.html)_

Changes:

* [x] Well-lnown Symbols
* [x] Loader Objects (loader base class definition)
* [x] Registry Objects (internal class definition for loader base class' [[Registry]] internal slot)
* [x] Module Objects (base class definition for Modules, former Reflective Module)
* [x] Browser Loader Objects (loader class for default loader instances implemented by browsers)
* [x] `System.loader` is an instance of Browser Loader
* [x] SitePackage was deferred (removed from the spec for now)
* [x] Use symbols to access hooks
* [x] Iterate over Registry (e.g.: `for (let entry of System.loader.registry) {}`)
* [x] Avoid leaking internal temporary slots from registry entries. (more work to do here, for now, we expose `state` and `statePromise` instead of exposing all available internal promises for every past state)
* [x] Reflective Module -> Module Objects (Reflect.Module)
* [x] Backpointer from registry to the loader instance it belongs to.
* [x] Reflect.Loader.prototype.load() is now name and referrer driven.
* [x] annexes

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -66,6 +66,88 @@ The primary goal is to make as much of this process as possible consistent betwe
 
 <h2 id="conventions">Conventions</h2>
 
+<h3 id="well-known-intrinsic-objects">Well-Known Intrinsic Objects</h3>
+
+Well-known intrinsics are built-in objects that are explicitly referenced by the algorithms of this specification and which usually have Realm specific identities. Unless otherwise specified each intrinsic object actually corresponds to a set of similar objects, one per Realm.
+
+Within this specification a reference such as %name% means the intrinsic object, associated with the current Realm, corresponding to the name. Determination of the current Realm and its intrinsics is described in ES2015, 8.3. The well-known intrinsics are listed in Table X.
```
Comment (dherman) [inline index.bs]: italicize *name*, as in:

"a reference such as %*name*% means the intrinsic object, associated with the current Realm, corresponding to *name*."

File: index.bs [hunk 2]
```diff
@@ -66,6 +66,88 @@ The primary goal is to make as much of this process as possible consistent betwe
...
+
+<table>
```
Comment (dherman) [inline index.bs]: The conventions section is only establishing spec conventions used in the rest of the spec, but it doesn't contain actual spec contents. So I wouldn't put this table here.

Since this table really cuts across different categories (especially across things that are universal vs browser-specific), I would put it in an appendix at the end, with a cell that refers to the section of the spec that introduces that row's intrinsic.

File: index.bs [hunk 3]
```diff
@@ -66,6 +66,88 @@ The primary goal is to make as much of this process as possible consistent betwe
...
+  <thead>
+    <tr>
+      <th>Intrinsic Name</th>
+      <th>Global Name</th>
+      <th>Description (<em>non-normative</em>)</th>
+    </tr>
+  </thead>
+  <tr>
```
Comment (dherman) [inline index.bs]: I don't believe we need a %Loader% or %LoaderPrototype% intrinsic. `Reflect.Loader` and `Reflect.Loader.prototype` should not have any particular powers or connection to the realm, and nothing in the semantics needs to refer to them; they're just regular library functions like, say, `parseInt`.

File: index.bs [hunk 4]
```diff
@@ -66,6 +66,88 @@ The primary goal is to make as much of this process as possible consistent betwe
...
+    <td>%BrowserLoader%</td>
+    <td></td>
+    <td>The browser loader constructor (<a href="#browser-loader">...</a>)</td>
+  </tr>
+  <tr>
+    <td>%BrowserLoaderPrototype%</td>
+    <td></td>
+    <td>The initial value of the *prototype* data property of %BrowserLoader%</td>
+  </tr>
+  <tr>
```
Comment (dherman) [inline index.bs]: Given that site-packages are half-baked right now, and slated for the latest stage of the [roadmap](https://github.com/whatwg/loader/blob/master/roadmap.md), I think we should just eliminate the feature completely from the spec for now.

Comment (caridy) [inline index.bs]: ok, will do.

File: index.bs (hunk 3 above)
Comment (caridy) [inline index.bs]: Got it, I will update the PR. As today, they are only used in BrowserLoader Objects section, but as you mentioned, I can just use `Reflect.Loader` and `Reflect.Loader.prototype` directly.

File: index.bs (hunk 1 above)
Comment (caridy) [inline index.bs]: ok

File: index.bs [hunk 5]
```diff
@@ -899,13 +1030,11 @@ The browser loader has an extra internal slot:
   <li>top-level packages consult \[[Site]]: <code>"jquery"</code>, <code>"ember/data"</code>
```
Comment (dherman) [inline index.bs]: delete this TODO since we're deferring site-packages

File: index.bs [hunk 6]
```diff
@@ -736,28 +881,15 @@ Reflective modules are always already instantiated.
...
-  <li>re-export (immutable): <code>{ module: m, import: "foo" }</code>
-</ul>
-
-<emu-alg>
-1. // TODO: spec me
-</emu-alg>
+The Module constructor is the initial value of the Module property of the the Reflect object. When called as a constructor it creates and initializes a new Module object. Reflect.Module is not intended to be called as a function and will throw an exception when called in that manner.
 
-<h4 id="create-module-mutator" aoid="CreateModuleMutator">CreateModuleMutator(module)</h4>
+The Reflect.Module constructor is designed to be subclassable. It may be used as the value in an extends clause of a class definition. Subclass constructors that intend to inherit the specified Module behaviour must include a super call to the Reflect.Module constructor to create and initialize the subclass instance with the internal state necessary to support the Reflect.Module.prototype built-in methods.
```
Comment (dherman) [inline index.bs]: change the end of this to "with the internal state necessary to integrated with loaders."

File: index.bs [hunk 7]
```diff
@@ -798,97 +930,103 @@ Reflective modules are always already instantiated.
...
+This property has the attributes { \[[Writable]]: false, \[[Enumerable]]: false, \[[Configurable]]: false }.
 
-The browser loader contains extra properties for storing <dfn>site packages</dfn>, an application-global set of globally available packages. These map in an internal table to unique URLs that in turn serve as keys in the module registry.
+<h3 id="module-prototype-object">Properties of the Reflect.Module Prototype Object</h3>
 
-<div class="note">
-<p>
-The site package system serves as a simple coordination mechanism for modest-sized applications, but it does not provide all functionality required of a full-fledged package management system. It is expected that development ecosystems will build around package management tools that deal with requirements outside the scope of this specification, such as version management and allowing multiple versions of a library to coexist with the same name.
-</p>
+<h4 id="Reflect.Module.prototype.constructor">Reflect.Module.prototype.constructor</h4>
```
Comment (dherman) [inline index.bs]: remove this -- there should not be a `constructor` property on the prototype (it can't have any properties).

File: index.bs [hunk 8]
```diff
@@ -798,97 +930,103 @@ Reflective modules are always already instantiated.
...
-<p>
-Tools that preprocess JavaScript source code may choose to use or ignore the site package table. For example, a package manager may choose to preprocess two separate import statements requiring <code>"jquery"</code> to <code>"jquery/1.9"</code> and <code>"jquery/2.1.1"</code> respectively, based on configuration files informing the tool of version requirements. The tool would then store both versions of jQuery in the site package table using the longer names. Alternatively, the tool may choose to preprocess the imports directly as URLs and bypass the site package system altogether.
-</p>
-</div>
+The initial value of Reflect.Module.prototype.constructor is Reflect.Module.
 
-The browser loader has an extra internal slot:
+<h3 id="reflect-module-internal-slots">Properties of Module Instances</h3>
+
+Reflect.Module instances are ordinary objects that inherit properties from the Reflect.Module.prototype.
```
Comment (dherman) [inline index.bs]: This isn't correct. Reflect.Module instances are "module namespace exotic objects" as defined in ES6. However, we should fix a slight design flaw in ES6, which is that in in 9.4.6.1, a module namespace exotic object always reports its prototype as null. We'll need to add a section in this spec describing the fix. We can do this in a separate PR.

File: index.bs [hunk 9]
```diff
@@ -798,97 +930,103 @@ Reflective modules are always already instantiated.
...
+
+Reflect.Module instances are initially created with the internal slots described in the following table:
```
Comment (dherman) [inline index.bs]: This stuff shouldn't get moved out of the section on Reflective Module Records. The \[[LocalExports]], \[[IndirectExports]], and \[[Evaluate]] internal properties belong on a Reflective Module Record, which is encapsulated internally by the module namespace exotic object returned by the `Reflect.Module` constructor.

**Comment by dherman** [comment]
So much amazing work. Thanks @caridy, this is such great progress. <3

**Comment by guybedford** [comment]
:+1:
